var/spack/repos/builtin/packages/chapel/fix_spack_cc_wrapper_in_cray_prgenv.patch

```diff
@@ -0,0 +1,82 @@
+     -L* | -R* | -l*)
+        # Some compilers place space between "-{L,R}" and the path.
+        # Remove the space.
+-       if test x-L = "$p" ||
+-          test x-R = "$p"; then
++       if test x-L = x"$p" ||
++          test x-R = x"$p" || test x-l = x"$p"; then
+ 	 prev=$p
+ 	 continue
+        fi
+diff --git a/third-party/hwloc/hwloc-src/configure b/third-party/hwloc/hwloc-src/configure
+index f918598357..51d728f144 100755
+--- a/third-party/hwloc/hwloc-src/configure
++++ b/third-party/hwloc/hwloc-src/configure
+@@ -16640,8 +16640,8 @@ if { { eval echo "\"\$as_me\":${as_lineno-$LINENO}: \"$ac_compile\""; } >&5
+     -L* | -R* | -l*)
+        # Some compilers place space between "-{L,R}" and the path.
+        # Remove the space.
+-       if test x-L = "$p" ||
+-          test x-R = "$p"; then
++       if test x-L = x"$p" ||
++          test x-R = x"$p" || test x-l = x"$p"; then
+ 	 prev=$p
+ 	 continue
+        fi
+diff --git a/third-party/libunwind/libunwind-src/configure b/third-party/libunwind/libunwind-src/configure
+index 164431c2db..d37a32497f 100755
+--- a/third-party/libunwind/libunwind-src/configure
++++ b/third-party/libunwind/libunwind-src/configure
+@@ -14751,8 +14751,8 @@ if { { eval echo "\"\$as_me\":${as_lineno-$LINENO}: \"$ac_compile\""; } >&5
+     -L* | -R* | -l*)
+        # Some compilers place space between "-{L,R}" and the path.
+        # Remove the space.
+-       if test x-L = "$p" ||
+-          test x-R = "$p"; then
++       if test x-L = x"$p" ||
++          test x-R = x"$p" || test x-l = x"$p"; then
+ 	 prev=$p
+ 	 continue
+        fi
+diff --git a/third-party/qthread/qthread-src/configure b/third-party/qthread/qthread-src/configure
+index 5e87b10852..5bad78d5e1 100755
+--- a/third-party/qthread/qthread-src/configure
++++ b/third-party/qthread/qthread-src/configure
+@@ -19780,8 +19780,8 @@ if { { eval echo "\"\$as_me\":${as_lineno-$LINENO}: \"$ac_compile\""; } >&5
+     -L* | -R* | -l*)
+        # Some compilers place space between "-{L,R}" and the path.
+        # Remove the space.
+-       if test x-L = "$p" ||
+-          test x-R = "$p"; then
++       if test x-L = x"$p" ||
++          test x-R = x"$p" || test x-l = x"$p"; then
+ 	 prev=$p
+ 	 continue
+        fi
+diff --git a/util/chplenv/chpl_llvm.py b/util/chplenv/chpl_llvm.py
+index f0fd495f28..95dc9c3f67 100755
+--- a/util/chplenv/chpl_llvm.py
++++ b/util/chplenv/chpl_llvm.py
+@@ -866,13 +866,14 @@ def get_clang_prgenv_args():
+         os.environ['PE_CHAPEL_PKGCONFIG_LIBS'] = gather_pe_chpl_pkgconfig_libs()
+
+         # Use cc --cray-print-opts=... to get arguments from compiler driver
+-
++        # find the actual cc in case something like spack has wrapped it
++        real_cc = os.path.join(os.environ["CRAYPE_DIR"], "bin", "cc")
+         # Get compilation arguments
+-        opts = run_command(['cc', '--cray-print-opts=cflags'])
++        opts = run_command([real_cc, '--cray-print-opts=cflags'])
```

Comment:
@scheibelp I don't really understand this feedback. 

Your comment is on a patch that fixes a defect in the Chapel's underlying configure logic, which currently makes the questionable assumption that the compiler used on a Cray EX is always the Cray PrgEnv wrapper. This is almost always the case for manual builds of Chapel, but Spack's compiler handling breaks this assumption.

This patch has (almost) nothing to do with RPATH or LD_LIBRARY_PATH: it's about avoiding a hard error in the Chapel build when `$CC` is actually a Spack compiler wrapper that rejects the special options provided by the Cray PrgEnv wrapper.

Reply:
It looked like this was replacing a call like `cc ...` with `/abspath/to/cc`, which I thought would bypass Spack's compiler wrapper. I didn't know what context this runs in (e.g. during the build, or after), but this could in particular be an issue if run during the build.

I noticed other parts of the PR setting LD_LIBRARY_PATH: that should not be required if the Spack compiler wrappers are used because they insert RPATHs. This was the only part of the PR that I saw directly interfering with compiler paths, but on closer examination it only seems to be changing that path for this single call, and not for other calls (e.g. when running `cc` to build objects).

>  it's about avoiding a hard error in the Chapel build when $CC is actually a Spack compiler wrapper that rejects the special options provided by the Cray PrgEnv wrapper.

based on this it sounds like I was mistaken and generally the build is using the Spack compiler wrappers (which is good, although that rekindles my confusion over spots that set `LD_LIBRARY_PATH`). The Spack compiler wrapper *should* be willing to pass arbitrary options through to the underlying compiler, so

* this seems like a bug in Spack's compiler wrapper
* I'm curious what error you got when running this without the patch

Reply:
> This was the only part of the PR that I saw directly interfering with compiler paths, ...
> generally the build is using the Spack compiler wrappers (which is good, although that rekindles my confusion over spots that set LD_LIBRARY_PATH).

This is not entirely true. A recent commit (which might post-date your reviewed version) explicitly bypasses the Spack compiler wrapper for the target compiler, because it needs to be usable in some cases when the Chapel compiler is invoked by the end-user after Spack installation. 

However I suspect this patch is still necessary to ensure the Host compiler (which remains the Spack wrapper) is handled correctly by Chapel configure.

> > it's about avoiding a hard error in the Chapel build when $CC is actually a Spack compiler wrapper that rejects the special options provided by the Cray PrgEnv wrapper.
> 
> based on this it sounds like I was mistaken and generally the build is using the Spack compiler wrappers (which is good, although that rekindles my confusion over spots that set `LD_LIBRARY_PATH`). The Spack compiler wrapper _should_ be willing to pass arbitrary options through to the underlying compiler, so
> 
> * this seems like a bug in Spack's compiler wrapper
> * I'm curious what error you got when running this without the patch

It's not that Spack's compiler wrapper is failing to pass the option, it's that Spack's compiler wrapper happily passes it to the underlying compiler (e.g. raw gcc) which is not Cray's compiler wrapper that unmodified Chapel configure expects on a Cray system. This leads to a failure inside the Chapel build that looks like:
```
gcc: error: unrecognized command-line option '--cray-print-opts=cflags'
```

IOW if there's a bug it's that Chapel assumes the C compiler on a Cray is always the Cray PrgEnv wrapper, and Spack automatically "unwraps" the Cray PrgEnv wrapper and targets the underlying native compiler directly (but I'm pretty sure that's by design).